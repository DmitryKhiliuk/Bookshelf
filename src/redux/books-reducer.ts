import {v1} from "uuid";

export type BookType = {
    id: string
    title: string
    author: string
    year: string
    image: string
}

export type ActionType = addBookACType | removeBookACType | editBookACType

const initialState: BookType[] = [
    {id:v1(), title: 'React: современные шаблоны для разработки ', author: 'Алекс Бэнкс, Ева Порселло', year: '2021', image: 'https://s4-goods.ozstatic.by/480/537/52/101/101052537_0_React_sovremennie_shabloni_dlya_razrabotki_prilozheniy_Aleks_Benks_Eva_Porsello.jpg'},
    {id:v1(), title: 'JavaScript с нуля', author: 'Кирупа Чиннатамби', year: '2021', image: 'https://s2-goods.ozstatic.by/480/443/28/101/101028443_0_JavaScript_s_nulya_Kirupa_CHinnatambi.jpg'},
    {id:v1(), title: 'Turbo Pascal', author: 'Валерий Фаронов', year: '2021', image: 'https://s3-goods.ozstatic.by/480/884/110/101/101110884_0_Turbo_Pascal_Uchebnoe_posobie_Valeriy_Faronov.jpg'},
    {id:v1(), title: 'Биполярное расстройство', author: 'Е. Касьянов', year: '2022', image: 'https://s2-goods.ozstatic.by/480/803/109/101/101109803_0_Bipolyarnoe_rasstroystvo_gid_po_vizhivaniyu_dlya_teh_kto_chasto_ne_vidit_beloy_polosi_Masha_Pushkina_E_Kasyanov.jpg'},
]

export const booksReducer = (state:BookType[] = initialState, action: ActionType):BookType[] => {
    switch (action.type) {
        case "ADD-BOOK": {
            return [action.newBook, ...state]
        }
        case "REMOVE-BOOK": {
            return state.filter((b) => b.id !== action.id)
        }
        case "EDIT-BOOK": {
            return state.map((b) => b.id === action.id ? {...b, ...action.newBook} : b)
        }
    }
    return state
}

export type addBookACType = ReturnType<typeof addBookAC>
export const addBookAC = (newBook:BookType) => {
    return {
        type: 'ADD-BOOK',
        newBook
    } as const
}
export type removeBookACType = ReturnType<typeof removeBookAC>
export const removeBookAC = (id:string) => {
    return {
        type: 'REMOVE-BOOK',
        id
    } as const
}
export type editBookACType = ReturnType<typeof editBookAC>
export const editBookAC = (id:string, newBook:BookType) => {
    return {
        type: 'EDIT-BOOK',
        id,
        newBook
    } as const
}