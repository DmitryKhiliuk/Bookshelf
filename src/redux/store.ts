import {combineReducers, createStore} from "redux";
import {booksReducer} from "./books-reducer";
import {loadState, saveState} from "./localstorage-utils";

export const rootReducer = combineReducers({
    books: booksReducer
})

export const store = createStore(rootReducer, loadState())

store.subscribe(() => {
    saveState({
        books:store.getState().books
    })
})

export type AppRootStateType = ReturnType<typeof rootReducer>

// @ts-ignore
window.store = store;