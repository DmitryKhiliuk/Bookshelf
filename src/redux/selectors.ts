import {AppRootStateType} from "./store";

export const selectBooks = (state: AppRootStateType) => state.books