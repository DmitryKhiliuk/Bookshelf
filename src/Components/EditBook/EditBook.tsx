import React, {useState} from 'react';
import {Button} from "../Button";
import {NavLink, useParams} from "react-router-dom";
import {useDispatch, useSelector} from "react-redux";
import {BookType, editBookAC} from "../../redux/books-reducer";
import {Input} from "../Input";
import {AppRootStateType} from "../../redux/store";
import s from './EditBook.module.css'
import {selectBooks} from "../../redux/selectors";

type EditBookType = {
    background: {backgroundImage: string}
}

export const EditBook = (props: EditBookType) => {

    const dispatch = useDispatch()
    const {bookId} = useParams()
    const books = useSelector<AppRootStateType, BookType[]>(selectBooks)
    console.log('bookId: ', bookId)
    const currentBook = books.find((book) => book.id === bookId )

    const[title, setTitle] = useState(currentBook && currentBook.title)
    const[author, setAuthor] = useState(currentBook && currentBook.author)
    const[year, setYear] = useState(currentBook && currentBook.year)
    const[image, setImage] = useState(currentBook && currentBook.image)

    let newBook: any  = {

    }

    const onClickSave = () => {
        newBook.id = bookId
        newBook.title = title
        newBook.author = author
        newBook.year = year
        newBook.image = image
        dispatch(editBookAC(newBook.id, newBook))

    }


    const onClickCancel = () => {
        console.log('edit book')
    }


    return (
        <div className={s.editBook} style={props.background}>
            <div className={s.back}>
                <h2 className={s.editBook__alignment}>Редактирование книги</h2>
                <div className={s.editBook__alignment}>
                    <div className={s.editBook__item__title}>Наименование</div>
                    <Input callBack={setTitle} value={title}/>
                </div>
                <div className={s.editBook__alignment}>
                    <div className={s.editBook__item__title}>Автор</div>
                    <Input callBack={setAuthor} value={author}/>
                </div>
                <div className={s.editBook__alignment}>
                    <div className={s.editBook__item__title}>Год выпуска</div>
                    <Input callBack={setYear} value={year}/>
                </div>
                <div className={s.editBook__alignment}>
                    <div className={s.editBook__item__title}>Изображение</div>
                    <Input callBack={setImage} value={image}/>
                </div>
            </div>
            {/*<div>
                {currentBook && JSON.stringify(currentBook, null, 2)}
            </div>*/}
            <div className={s.editBook__alignment}>
                <div className={s.editBook__button}><NavLink to={'/'}><Button callBack={onClickSave} name={'Сохранить'}/></NavLink></div>
                <div className={s.editBook__button}><NavLink to={'/'}><Button callBack={onClickCancel} name={'Отменить'}/></NavLink></div>
            </div>
        </div>
    );
};

