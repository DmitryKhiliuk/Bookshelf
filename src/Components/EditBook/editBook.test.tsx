import React from "react";
import {render, screen} from "@testing-library/react";
import {MemoryRouter} from "react-router-dom";
import {EditBook} from "./EditBook";
import backgroundImage from "../../assets/2423.jpg";
import * as reduxHooks from "react-redux";
import userEvent from "@testing-library/user-event";
import {AppRootStateType} from "../../redux/store";

const background = {backgroundImage: `url(${backgroundImage} )`}

jest.mock('react-redux')
const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch')
const mockedSelector = jest.spyOn(reduxHooks, 'useSelector')
const dispatch = jest.fn()

jest.mock('console')
const mockedLog = jest.spyOn(console, 'log')

let startState: AppRootStateType

describe('edit book component test', () => {
    beforeEach(() => {
        startState = {
            books: [
                {
                    id: '1',
                    title: "Book",
                    author: "writer",
                    year: "1900",
                    image: "url",
                },
                {
                    id: '2',
                    title: "Book2",
                    author: "writer2",
                    year: "1902",
                    image: "url2",
                },

            ]
        }
        mockedSelector.mockReturnValue(startState.books)
    })
    it('should create edit book component', () => {
        const view = render(<MemoryRouter><EditBook background={background}/></MemoryRouter>)
        expect(view).toMatchSnapshot()
    })
    it('must render buttons in the component edit book', () => {
        render(<MemoryRouter><EditBook background={background}/></MemoryRouter>)
        const saveBtn = screen.getByText('Сохранить')
        const cancelBtn = screen.getByText('Отменить')
        const allBtn = screen.getAllByRole('button')
        expect(allBtn.length).toEqual(2)
        expect(allBtn).toEqual([saveBtn, cancelBtn])
    })
    it('must render inputs in the component edit book', () => {
        render(<MemoryRouter><EditBook background={background}/></MemoryRouter>)
        const allInputs = screen.getAllByRole('textbox')
        expect(allInputs.length).toEqual(4)
    })
    it('should dispatch edit book action', () => {
        mockedDispatch.mockReturnValue(dispatch)
        render(<MemoryRouter><EditBook background={background}/></MemoryRouter>)
        userEvent.click(screen.getByText('Сохранить'))
        expect(dispatch).toHaveBeenCalledTimes(1)

    })
    it('main page button test', () => {
        mockedLog.mockReturnValue()
        render(<MemoryRouter><EditBook background={background}/></MemoryRouter>)
        userEvent.click(screen.getByText('Отменить'))
        expect(mockedLog).toHaveBeenCalled()
    })
})