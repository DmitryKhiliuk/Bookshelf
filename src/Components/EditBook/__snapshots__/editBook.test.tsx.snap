// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit book component test should create edit book component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="editBook"
      >
        <div
          class="back"
        >
          <h2
            class="editBook__alignment"
          >
            Редактирование книги
          </h2>
          <div
            class="editBook__alignment"
          >
            <div
              class="editBook__item__title"
            >
              Наименование
            </div>
            <div>
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="editBook__alignment"
          >
            <div
              class="editBook__item__title"
            >
              Автор
            </div>
            <div>
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="editBook__alignment"
          >
            <div
              class="editBook__item__title"
            >
              Год выпуска
            </div>
            <div>
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="editBook__alignment"
          >
            <div
              class="editBook__item__title"
            >
              Изображение
            </div>
            <div>
              <input
                class="input"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="editBook__alignment"
        >
          <div
            class="editBook__button"
          >
            <a
              class=""
              href="/Bookshelf"
            >
              <div>
                <button
                  class="button"
                >
                  Сохранить
                </button>
              </div>
            </a>
          </div>
          <div
            class="editBook__button"
          >
            <a
              class=""
              href="/Bookshelf"
            >
              <div>
                <button
                  class="button"
                >
                  Отменить
                </button>
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="editBook"
    >
      <div
        class="back"
      >
        <h2
          class="editBook__alignment"
        >
          Редактирование книги
        </h2>
        <div
          class="editBook__alignment"
        >
          <div
            class="editBook__item__title"
          >
            Наименование
          </div>
          <div>
            <input
              class="input"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="editBook__alignment"
        >
          <div
            class="editBook__item__title"
          >
            Автор
          </div>
          <div>
            <input
              class="input"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="editBook__alignment"
        >
          <div
            class="editBook__item__title"
          >
            Год выпуска
          </div>
          <div>
            <input
              class="input"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="editBook__alignment"
        >
          <div
            class="editBook__item__title"
          >
            Изображение
          </div>
          <div>
            <input
              class="input"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="editBook__alignment"
      >
        <div
          class="editBook__button"
        >
          <a
            class=""
            href="/Bookshelf"
          >
            <div>
              <button
                class="button"
              >
                Сохранить
              </button>
            </div>
          </a>
        </div>
        <div
          class="editBook__button"
        >
          <a
            class=""
            href="/Bookshelf"
          >
            <div>
              <button
                class="button"
              >
                Отменить
              </button>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
